### DISCUSSION TITLE
Slides from session 3: Customizing your RAG Chat App

### DISCUSSION THREAD

**Discussion by multispark** [microsoft/AI-Chat-App-Hack#36] [Announcements]
Slides: https://speakerdeck.com/pamelafox/customizing-your-rag-chat-app
Video: https://www.youtube.com/watch?v=vt7oZg4bPAQ

Links from presentation:

* [Guide to free deployment](https://aka.ms/ragchat-free)
* [Data ingestion guide: Removing documents](https://github.com/pamelafox/azure-search-openai-demo/blob/main/docs/data_ingestion.md)
* [Data ingestion guide: Adding documents](https://github.com/Azure-samples/azure-search-openai-demo/blob/main/docs/data_ingestion.md)
* [Converting HTML pages with Playwright](https://blog.pamelafox.org/2024/01/converting-html-pages-to-pdfs-with.html)
* [PR: Adding a JSONParser class](https://github.com/Azure-Samples/azure-search-openai-demo/pull/1195)
* [Integrated data chunking and embedding in Azure AI Search](https://learn.microsoft.com/azure/search/vector-search-integrated-vectorization)

**Comment by jmc9000** [comment]
Links Chris shared:

- [Chat with your Azure DevOps data](https://techcommunity.microsoft.com/t5/fasttrack-for-azure/chat-with-your-azure-devops-data/ba-p/4017784)
- [AI-in-a-Box](https://github.com/Azure/AI-in-a-Box)

**Comment by pamelafox** [comment]
Additional links shared by me:

* [Pamela's blog posts on OpenAI](https://blog.pamelafox.org/search/label/openai)
* [PR: Add Markdown support](https://github.com/Azure-Samples/azure-search-openai-demo/pull/56/files)
* [HTMLParser class from other repo](https://github.com/microsoft/sample-app-aoai-chatGPT/blob/1e401acec3a713b528a1b9b83ee6e70fce10e208/scripts/data_utils.py#L299)

**Comment by Thukyd** [comment]
I've got trouble to get this running on my local machine. I am able to start the backend server and frontend. But when I open the frontend url, I get errors as below. I found some proxy configs in the vite.config.ts but don't know how to further debug this. 

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/7248925/b997fc81-5f01-4ab6-8233-f5700ca06dcc)


![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/7248925/124144b3-4317-4aed-a901-16b328492ca7)

![image](https://github.com/microsoft/AI-Chat-App-Hack/assets/7248925/c6dd41f8-5472-4352-a6c9-d2b8b176d771)

**Reply by itamitCom** [reply]
I had the same problem

replace localhost with 127.0.0.1

you should get it like this

<img width="562" alt="image" src="https://github.com/microsoft/AI-Chat-App-Hack/assets/8110064/7ebfe70d-7a53-47dc-bd9b-76ed39541393">

**Reply by Thukyd** [reply]
Thanks a million @itamitCom, the page load now!

On server side a I get a SSL-Cert error but still got to check further. 

`A body is sent with the request
ERROR:root:Exception while generating response stream: Cannot connect to host gptkb-ambj53mwk6u5g.search.windows.net:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1002)')]
Traceback (most recent call last):
  File "/Users/jp/Documents/GitHub/azure-search-openai-hackathon/app/backend/backend_env/lib/python3.11/site-packages/aiohttp/connector.py", line 992, in _wrap_create_connection
    return a`

**Reply by pamelafox** [reply]
Certificate issues often relate to a malformed Python env, you can try the suggestions here: https://stackoverflow.com/questions/35569042/ssl-certificate-verify-failed-with-python3/43855394#43855394

**Comment by edifice1989** [comment]
hi @pamelafox, I encountered an issue while trying to deploy locally. I have no issues previously deploying to Azure tho.
below is the error message. All Azure keys and Openai keys seem right. App is up and running, the error only happens when I ask the question and generating answers. I could not debug the issue any further, any suggestions? thx in advance!

ERROR:root:Exception while generating response stream: () Authorization failed.
Code:
Message: Authorization failed.
Traceback (most recent call last):
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/app.py", line 150, in format_as_ndjson
    async for event in r:
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/approaches/chatapproach.py", line 149, in run_with_streaming
    extra_info, chat_coroutine = await self.run_until_final_call(
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/approaches/chatreadretrieveread.py", line 152, in run_until_final_call
    results = await self.search(top, query_text, filter, vectors, use_semantic_ranker, use_semantic_captions)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/approaches/approach.py", line 136, in search
    async for page in results.by_page():
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/backend_env/lib/python3.11/site-packages/azure/core/async_paging.py", line 94, in __anext__
    self._response = await self._get_next(self.continuation_token)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/backend_env/lib/python3.11/site-packages/azure/search/documents/aio/_paging.py", line 107, in _get_next_cb
    return await self._client.documents.search_post(search_request=self._initial_query.request, **self._kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/backend_env/lib/python3.11/site-packages/azure/core/tracing/decorator_async.py", line 88, in wrapper_use_tracer
    return await func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/c/Users/Ming/Desktop/chatbot/azure-search-openai-demo/app/backend/backend_env/lib/python3.11/site-packages/azure/search/documents/_generated/aio/operations/_documents_operations.py", line 417, in search_post
    raise HttpResponseError(response=response, model=error)
azure.core.exceptions.HttpResponseError: () Authorization failed.

**Reply by pamelafox** [reply]
Hm, well that error indicates that your local credentials aren't correctly authenticating to the search service. You say you're logged in correctly. I wonder if you are logged into a different account than the one you deployed with? When you ran "azd up", we gave permissions for the current logged in user to use the search service, so if you change accounts, those permissions will no longer work. We do have a script at https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/scripts/roles.sh that will setup those same permissions for the local user that you could try.

**Reply by edifice1989** [reply]
I got error when I run roles.sh
usage error: --assignee STRING | --assignee-object-id GUID
usage error: --assignee STRING | --assignee-object-id GUID
usage error: --assignee STRING | --assignee-object-id GUID
usage error: --assignee STRING | --assignee-object-id GUID
usage error: --assignee STRING | --assignee-object-id GUID

**Reply by pamelafox** [reply]
Sorry, it's missing AZURE_PRINCIPAL_ID. Full steps for running are here: https://github.com/Azure-Samples/azure-search-openai-demo/?tab=readme-ov-file#sharing-environments

**Reply by edifice1989** [reply]
that works for me: ) thank you so much!
